arm-software/shared/elf2bin/Opts.td

```diff
@@ -0,0 +1,64 @@
+                   HelpText<help>,
+                   MetaVarName<metavar>,
+                   Group<group>;
+  def : Separate<["--"], longname>,
+        Alias<!cast<Joined>(NAME # "_EQ")>;
+
+  if !not(!empty(shortname)) then {
+    def : JoinedOrSeparate<["-"], shortname>,
+          Alias<!cast<Joined>(NAME # "_EQ")>,
+          HelpText<"Alias for --" # longname>, Group<group>;
+  }
+}
+
+multiclass NoValue<string longname, string shortname, string help,
+                   OptionGroup group> {
+  def NAME : Flag<["--"], longname>,
+             HelpText<help>,
+             Group<group>;
+
+  if !not(!empty(shortname)) then {
+    def : Flag<["-"], shortname>,
+          Alias<!cast<Flag>(NAME)>,
+          HelpText<"Alias for --" # longname>;
+  }
+}
+
+def grp_mode: OptionGroup<"mode">, HelpText<"FORMAT OF OUTPUT">;
+def grp_file: OptionGroup<"output">, HelpText<"LOCATION OF OUTPUT">;
+def grp_opts: OptionGroup<"options">, HelpText<"OPTIONS">;
+
+defm output_file: Value<"output-file", "o", "FILENAME", "Name of the output file. Invalid if more than one output file is to be generated.", grp_file>;
+defm output_pattern: Value<"output-pattern", "O", "PATTERN", "Schema for naming all output files. May contain %f for base name of input file; %F for full name of input file; %a or %A for base address of ELF segment, in hex or in HEX; %b for bank number, if using --banks; %% for a literal % sign.", grp_file>;
+
+defm ihex: NoValue<"ihex", "", "Output Intel Hex files (\":108000\" style)", grp_mode>;
+defm srec: NoValue<"srec", "", "Output Motorola Hex files (\"S3150800\" style)", grp_mode>;
+defm bin: NoValue<"bin", "", "Output a binary file per ELF segment", grp_mode>;
+defm bincombined: NoValue<"bincombined", "", "Output a single binary file, containing all segments, with padding to place them at the correct relative positions", grp_mode>;
+defm vhx: NoValue<"vhx", "", "Output a Verilog Hex file per ELF segment", grp_mode>;
+defm vhxcombined: NoValue<"vhxcombined", "", "Output a single Verilog Hex file, containing all segments, with padding to place them at the correct relative positions", grp_mode>;
+
+defm base: Value<"base", "", "ADDRESS", "Base address of the whole output file, for --bincombined or --vhxcombined", grp_opts>;
+defm banks: Value<"banks", "", "WIDTHxNUM", "Partition the output into subfiles by writing WIDTH bytes in turn to each of NUM files in cyclic order, for binary or Verilog Hex output", grp_opts>;
+defm datareclen: Value<"datareclen", "", "LENGTH", "Number of data bytes to put into each record, for Intel Hex or Motorola Hex output", grp_opts>;
+defm segments: Value<"segments", "", "ADDRESS,ADDRESS,...", "Select only the ELF segments of the input which start at the specified addresses", grp_opts>;
+defm zi: NoValue<"zi", "", "Include the zero-initialized padding at the end of each ELF segment", grp_opts>;
+defm physical: NoValue<"physical", "", "Use the physical addresses (p_paddr) of ELF segments", grp_opts>;
```

Comment:
Could it be worth a single option that takes physical or virtual as a parameter rather than two options. I'm sure we don't support both at the same time.

Reply:
Indeed not, but I think any other approach involves more typing: `--no-virtual` to contradict `--virtual`, or `--address=physical` vs `--address=virtual`. Brevity is useful.